src/mlpack/methods/ann/activation_functions/softshrink_function.hpp

```diff
@@ -0,0 +1,101 @@
+/**
+ * @file softshrink_function.hpp
+ * @author Lakshya Ojha
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_ANN_ACTIVATION_FUNCTIONS_SOFT_SHRINK_FUNCTION_HPP
+#define MLPACK_METHODS_ANN_ACTIVATION_FUNCTIONS_SOFT_SHRINK_FUNCTION_HPP
+
+#include <mlpack/prereqs.hpp>
+#include <algorithm>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * SoftShrinkage operator is defined as,
+ *         | x - lambda, if x >  lambda
+ *  f(x) = | x + lambda, if x < -lambda
+ *         | 0, otherwise
+ *
+ * lambda is set to 0.5 by default.
+ *
+ *          | 1 , if x >  lambda
+ *  f'(x) = | 1 , if x < -lambda
+ *          | 0 , otherwise
+ */
+class SoftShrinkFunction
+{
+ public:
+  /**
+   * Computes the Soft Shrinkage function.
+   *
+   * @param x Input data.
+   * @return f(x).
+   */
+  static double Fn(const double x, const double lambda = 0.5)
+  {
+    if (x > lambda)
+    {
+      return x - lambda;
+    }
+    else if (x < -1 * lambda)
+    {
+      return x + lambda;
+    }
+    return 0.;
+  }
+
+  /**
+   * Computes the Soft Shrinkage function.
+   *
+   * @param x Input data.
+   * @param y The resulting output activation.
+   */
+  template <typename InputVecType, typename OutputVecType>
+  static void Fn(const InputVecType &x, OutputVecType &y,
+                  const double lambda = 0.5)
+  {
+    y = x;
```

Comment:
Y.set_size(arma::size(x));

Reply:
Same in deriv too

Reply:
Hi @jeffin143 if we use
`y.set_size(arma::size(x));`
 then we will have to also use  `y = x;`   in the next line because we need y to be exactly same as x before we use the **transform** function on y. Otherwise it will be implemented on a y which has random values instead of our required ones. That's why its better to have it all done in a single line.
This is my hypothesis please verify the same and let me know the result.
By the way thank you for pointing out. It's always good to try new things.
:)

Reply:
I think that's ok.

Reply:
Just to mention the transform function is just below this `y = x;` i.e, line no 65.
`y.transform([](const double x, const double lambda = 0.5){
                  return Fn(x, lambda);});`
Thank You.

Reply:
> I think that's ok. However a member of this organization will provide better insight.
> Thanks.

Yes

Reply:
What I thought was 
y=x would copy a million items and then
y.transform would again make million iterations

but 
if you set size
and then a run a loop across a million times it would be a million less instruction

But I guess the compiler would optimise both the instruction, and hence there wouldn't be any difference

Reply:
Hi @jeffin143 the reason I didn't use 
`y.set_size(arma::size(x));` 
is that it only sets the size of y equal to x and doesn't initiate it to any values so it will have random values which also will take time as these are not done in O(1). Hence to set all values of y equal to x we would again have to use 
`y = x;` which I believe will again take a million iterations to set all values of y to x, which is a waste.
So it's better if we stick with `y = x;`
Hope this clears up :)